lib/Mir.ml

```diff
@@ -101,35 +84,48 @@ and 's statement =
    We'll need to scan the list of declarations for top-level ones and essentially remove them
    from the block. The AST has an `is_global` flag that also tracks this.
 *)
-type top_var_decl =
-  {tvident: string; tvtype: sizedtype; tvtrans: transformation; tvloc: string}
+type 'e top_var_decl =
+  { tvident: string
+  ; tvtype: 'e Ast.sizedtype
+  ; tvtrans: 'e Ast.transformation
+  ; tvloc: Ast.location_span sexp_opaque }
 [@@deriving sexp]
 
-type top_var_table = (string, top_var_decl) Map.Poly.t [@@deriving sexp]
+type 'e top_var_table = (string, 'e top_var_decl) Map.Poly.t [@@deriving sexp]
 
-type 's prog =
+type ('e, 's) prog =
   { functions_block: 's list
-  ; data_vars: top_var_table
-  ; tdata_vars: top_var_table
+  ; data_vars: 'e top_var_table
+  ; tdata_vars: 'e top_var_table
   ; prepare_data: 's list
-  ; params: top_var_table
-  ; tparams: top_var_table
+  ; params: 'e top_var_table
+  ; tparams: 'e top_var_table
   ; prepare_params:
       's list
       (* XXX too intimately tied up with stan reader.hpp and writer.hpp in codegen
      TODO: codegen parameter constraining and unconstraining in prepare_params
   *)
   ; log_prob: 's list
-  ; gen_quant_vars: top_var_table
+  ; gen_quant_vars: 'e top_var_table
   ; generate_quantities: 's list
   ; prog_name: string
   ; prog_path: string }
 [@@deriving sexp]
 
+type expr_typed_located =
```

Comment:
It might be useful to define a generic recursive record type for expression trees with metadata and statement trees with metadata, so that functions can be polymorphic across types of metadata. Something like:

```
type 'm expr_with = { expr: (expr_with 'm) expr; with: 'm }
type 'm stmt_with = { stmt: (stmt_with 'm) statement; with: 'm }
```

So then e.g. expr_typed_located could be written as
`expr_type_location expr_with`
where expr_type_location is a type that contains the type, loc, and adlevel. @VMatthijs I think you might have run into this too?

Reply:
Ahhhh yeah we were trying to figure out what to do about this!!!! We even had a long conversation about it. Does this solve that recursive problem we keep having where e.g. pp_statement used to not give a shit about the metadata but when I had to do something different for UDFs with a `Skip` body, that broke down: https://github.com/stan-dev/stanc3/blob/master/lib/Stan_math_code_gen.ml#L246

Reply:
I think it would, if I understand the issue

Reply:
+1. This should give you some limited sense of row polymorphism at the cost of one extra level of in direction in the meta data.

Reply:
Yeah, I don't love that extra level. And if we keep using records and just constellation of metadata, we can just keep adding and removing stuff and using the `; _}` pattern to avoid caring about the rest of the record. But on the other hand testing these functions would be much easier if they were polymorphic over metadata. But also even for `pp_expr` we need to know that we have the type of the expression for printing the indices, so I'm not sure any of them can be polymorphic over any metadata.

Reply:
Right. It does sound like this wouldn't solve the problem for expressions. That's a pity.

Reply:
I think it's probably worth spending some time playing around with it, so I will do that now and report back...